### PR TITLE
21 08 19

### PR DESCRIPTION
**Pull request by lindajohnstone**

### CODE REVIEW

File: MarsRover.Tests/MapParserShould.cs [hunk 1]
```diff
@@ -9,18 +9,15 @@ public class MapParserShould
...
         [Theory]
         [MemberData(nameof(MapParserShouldTestData.MapParserTestData), MemberType = typeof(MapParserShouldTestData))]
-        public void ReturnMap_GivenValidString(List<Square> squares, int width, int height, int expectedObstacleCount, int expectedSquareCount, params string[] inputs)
+        public void ReturnMap_GivenValidString(List<Square> squares, int width, int height, params string[] inputs)
         {
             var mapString = String.Join(Environment.NewLine, inputs);
             var expected = new Map(width, height, squares);
 
             var result = MapParser.ParseMap(mapString);
             var actualObstacleCount = result.Squares.FindAll(c => c.Content == SquareContent.Obstacle).Count;
```
Comment (indianbread) [inline MarsRover.Tests/MapParserShould.cs:18]: this can be removed

File: MarsRover.Tests/InputParserShould.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        [Theory]
+        [InlineData("N", Direction.North)]
+        [InlineData("n", Direction.North)]
+        [InlineData("S", Direction.South)]
+        [InlineData("s", Direction.South)]
+        [InlineData("E", Direction.East)]
+        [InlineData("e", Direction.East)]
+        [InlineData("W", Direction.West)]
+        [InlineData("w", Direction.West)]
+        public void ReturnRoverInitialDirection_GivenValidString(string input, Direction expected)
```
Comment (kedoodle) [inline MarsRover.Tests/InputParserShould.cs:18]: Great test. I think the name implies coupling with the Rover in some way, but it really just provides a Direction given a valid letter equivalent. The parser doesn't care what you do with the Direction (in this case, instantiation of a Rover)

File: MarsRover.Tests/InputParserShould.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            var result = InputParser.ParseDirection(input);
+
+            Assert.Equal(expected, result);
+        }
+
+        [Theory]
+        [InlineData("0,0", 0, 0)]
+        [InlineData("100,8", 100, 8)]
+        [InlineData("1,80", 1, 80)]
+        public void ReturnRoverInitialLocation_GivenValidString(string input, int x, int y)
```
Comment (kedoodle) [inline MarsRover.Tests/InputParserShould.cs:29]: Same here. From the perspective of the parser, it just gives you a Location, not necessarily a starting point for the Rover

File: MarsRover/FileInput.cs [hunk 4]
```diff
@@ -9,10 +9,10 @@ public string Read(string input)
             return File.ReadAllText(input);
         }
 
-        public string FileExists(string filePath)
+        public bool FileExists(string filePath)
         {
-            if (File.Exists(filePath)) return filePath;
-            return "File does not exist.";
+            if (File.Exists(filePath)) return true;
+            return false;
```
Comment (kedoodle) [inline MarsRover/FileInput.cs:15]: `return File.Exists(filePath);`

File: MarsRover/Validator.cs [hunk 5]
```diff
@@ -21,8 +21,8 @@ private static bool AreAllLinesEqual(string[] lines)
         {
             foreach (var line in lines)
             {
-                var linesAreEqual = lines.All(c => c.Length == line.Length);
-                if (!linesAreEqual) return false;
+                var linesAreEqualLength = lines.All(c => c.Length == line.Length);
```
Comment (kedoodle) [inline MarsRover/Validator.cs:24]: sorry didn't pick this up last time, but `c` should probably be `l` for line

File: MarsRover/FileInput.cs [hunk 6]
```diff
@@ -9,10 +9,10 @@ public string Read(string input)
             return File.ReadAllText(input);
         }
 
-        public string FileExists(string filePath)
+        public bool FileExists(string filePath)
         {
-            if (File.Exists(filePath)) return filePath;
-            return "File does not exist.";
+            if (File.Exists(filePath)) return true;
```
Comment (indianbread) [inline MarsRover/FileInput.cs:14]: you can just do

`return File.Exists(filePath)` as File.Exists returns a bool already. However, the filePath parameter is nullable (denoted by the question mark) so it's good practice to handle that too. https://docs.microsoft.com/en-us/dotnet/api/system.io.file.exists?view=net-5.0

File: MarsRover/Validator.cs [hunk 7]
```diff
@@ -31,15 +31,42 @@ private static bool AreAllCharsValid(string[] lines)
...
             {
-                var charsInElementAreValid = line.All(c => c == 'O' || c == 'N');
-                if (!charsInElementAreValid) return false;
+                var charsInLineAreValid = line.All(c => c == 'O' || c == 'N');
+                if (!charsInLineAreValid) return false;
             }
             return true;
         }
 
+        public static bool IsValidDirection(string input)
```
Comment (indianbread) [inline MarsRover/Validator.cs:40]: an alternative way of checking for a valid direction:

`var validDirections = new [] {"N", "S", "W", "E"};
return validDirections.Contains(input, StringComparison.InvariantCultureIgnoreCase);
`

File: MarsRover/Validator.cs [hunk 8]
```diff
@@ -31,15 +31,42 @@ private static bool AreAllCharsValid(string[] lines)
...
+            // split input on ','
+            // check if 2 elements
+            // try to parse first element
+            // check it is >= 0 & < width
+            // try to parse second element
+            // check it is >= 0 & < width
+            var coordinates = SplitInput(input, ",");
+            var hasTwoCoordinates = coordinates.Length == 2;
+            if (!hasTwoCoordinates) return false;
+            var xIsValid = Int32.TryParse(coordinates[0], out var x) && x >= 0 && x < width;
```
Comment (indianbread) [inline MarsRover/Validator.cs:60]: The checking for valid coordinates can be extracted to its own function

**Review by indianbread — APPROVED** [review]

**Pull request merged**